### PR TITLE
Adding tutorial content videos

### PR DESCRIPTION
**Pull request by jukent**
Adding cards for tutorial video recordings to resource gallery.

See:
https://github.com/ProjectPythia/projectpythia.github.io/issues/146
https://github.com/ProjectPythia/pythia-foundations/discussions/118

### CODE REVIEW

**Comment by jukent** [comment]
`blackdock` is failiing with a `ModuleNotFound` Error -- `toml`

"""
blackdoc.................................................................Failed
- hook id: blackdoc
- exit code: 1

Traceback (most recent call last):
  File "/home/runner/.cache/pre-commit/repoqsdzclid/py_env-python3/bin/blackdoc", line 5, in <module>
    from blackdoc.__main__ import main
  File "/home/runner/.cache/pre-commit/repoqsdzclid/py_env-python3/lib/python3.9/site-packages/blackdoc/__main__.py", line 12, in <module>
    from .blackcompat import (
  File "/home/runner/.cache/pre-commit/repoqsdzclid/py_env-python3/lib/python3.9/site-packages/blackdoc/blackcompat.py", line 9, in <module>
    import toml
ModuleNotFoundError: No module named 'toml'
"""

**Comment by keewis** [comment]
you can fix that by running `pre-commit autoupdate --repo blackdoc` (`v0.3.3` and below didn't declare `toml` as a explicit dependency)

**Comment by jukent** [comment]
> you can fix that by running `pre-commit autoupdate --repo blackdoc` (`v0.3.3` and below didn't declare `toml` as a explicit dependency)

Can you say more? I ran the above command on my local fork and nothing happened. I had to install `pre-commit` first.

**Comment by keewis** [comment]
The only thing it does is upgrade the version from `v0.3.3` to `v0.3.4`, which you can do manually by editing `.pre-commit-config.yaml` (sorry for making this more complicated than it really is)

**Review by kmpaul — COMMENTED** [review]
I don't understand why I don't see the new cards in the Resource Gallery.  ...?

@jukent

**Comment by jukent** [comment]
> I don't understand why I don't see the new cards in the Resource Gallery. ...?
> 
> @jukent

Me neither. I'll have to look into this more.

**Comment by jukent** [comment]
Fixed! Thanks @kmpaul

**Review by kmpaul — APPROVED** [review]
Fantastic!  Thanks, @jukent!

**Review by clyne — COMMENTED** [review]
Thanks for getting all of these together, @jukent!

We had a lot of discussion about this topic in our meeting, and I may have misunderstood how we were planning to move forward. But I thought the plan was to be able to showcase the Pythia (Xdev) created content, differentiating it from the external content. Given the limitation that the gallery can only filter by one item currently, I thought we had decided to create a unique tag for these tutorials (e.g. 'Xdev-videos') AND, perhaps more importantly, add a heading to our landing page that gave a short description (similar to what we currently have for "The Foundations Book" and "The Resource Gallery" with a link that would take you to the gallery with the 'Xdev-videos' tag (or whatever) selected. 

Secondly, we had also previously discussed providing links to the notebooks that go along with the videos and doing what we could to consolidate them so they not in peoples personnel repos. Moving the repos is probably a big lift, and maybe it is only realistic to try and ensure that moving forward new tutorials have a home somewhere under GitHub/ProjectPythia. But I'm wondering if for the purposes of this PR can we add the links to the repos to the respective gallery cards? Would that make sense?

cc @brian-rose for his input

**Comment by jukent** [comment]
Thanks @clyne What I took from that conversation was that 1) Yes we want to showcase the Pythia content in a unique way and 2) We don't want the resource gallery to exclude the Pythia content. This PR only addresses the second point by adding the videos to the resource gallery.  

PR #160 is a start at addressing how to get a page of Pythia-only content (but there is a conversation I'd like to have over there about Pythia vs Pangeo, and what affiliations should be included). There is more work to be done and conversations to be had with the landing page and search-ability.

Every video that has a repo, has the repo inked in the gallery. I can go ahead and add the links to the body of the card as well.

**Comment by jukent** [comment]
I've added the links and they work EXCEPT if the card has a "see more" because the description is too long. Then the html of that "see more" modal button seems to be affected somehow by the link to the content. This happens namely in the "Your First" Python Tutorial Seminar Series. I'm not sure why this is happening. @kmpaul

**Comment by clyne** [comment]
> Thanks @clyne What I took from that conversation was that 1) Yes we want to showcase the Pythia content in a unique way and 2) We don't want the resource gallery to exclude the Pythia content. This PR only addresses the second point by adding the videos to the resource gallery.

We're on the same page there!

> 
> PR #160 is a start at addressing how to get a page of Pythia-only content (but there is a conversation I'd like to have over there about Pythia vs Pangeo, and what affiliations should be included). There is more work to be done and conversations to be had with the landing page and search-ability.

Sounds good. Thanks for clarifying. 

> 
> Every video that has a repo, has the repo inked in the video. I can go ahead and add the links to the body of the card as well.

That would be great. Then I will approve! Thanks

**Comment by jukent** [comment]
Fixed the truncating of html issue (that was affecting the modal buttons) with this code from @eentzel https://github.com/eentzel/htmltruncate.py/blob/master/htmltruncate.py and help from @kmpaul :)

**Comment by jukent** [comment]
@clyne All of the tutorial video cards now have the link to the static content in their description! 
This had an unexpected hiccup that has been fixed with the new `_extensions/truncatehtml.py` file.

File: portal/_extensions/truncatehtml.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    Any tags that would be left open by truncation will be closed at
+    the end of the returned string.  Optionally append ellipsis if
+    the string was truncated."""
+    stack = []  # open tags are pushed on here, then popped when the matching close tag is found
+    retval = []  # string to be returned
+    length = 0  # number of characters (not counting markup) placed in retval so far
+    tokens = Tokenizer(str)
+    tok = tokens.next_token()
+    while tok != END:
+        if not length < target_len:
```
Comment (kmpaul) [inline portal/_extensions/truncatehtml.py:158]: Changing the termination condition to the following will truncate only on blank spaces "near" the target character length:

```suggestion
        if length >= target_len and tok == ' ':
```

**Review by kmpaul — CHANGES_REQUESTED** [review]
I figured out how to truncate on words (or really to cut on blank spaces).  If you implement the suggestion, it should work.

**Review by clyne — COMMENTED** [review]
Looks great, @jukent. One typo and one question:

On all of the cards "conent" should be "content"

Do we still have an open issue to provide a means to showcase [filter] all of the Pythia created videos so that we can, for example, add something to our main landing page that says something like "Recorded Project Pythia" tutorials, gives a brief description, and then provides a link to all of the Pythia created videos?

**Comment by jukent** [comment]
Good catch! Just fixed that now.

Yes I think creating that new page would be a great place idea. I can create the issue for it but am about to take some PTO until late October - so someone might want to move forward on that sooner than that.

**Review by clyne — APPROVED** [review]

**Comment by clyne** [comment]
> Good catch! Just fixed that now.
> 
> Yes I think creating that new page would be a great place idea. I can create the issue for it but am about to take some PTO until late October - so someone might want to move forward on that sooner than that.

Yes, it would be great if you can open an issue. Enjoy your PTO!

**Comment by jukent** [comment]
@clyne Do you think this is the same as https://github.com/ProjectPythia/projectpythia.github.io/issues/157 ?

**Comment by kmpaul** [comment]
I think it's the same, but you should make a note of what you are doing in that issue and cross-reference to this.

**Comment by clyne** [comment]
> @clyne Do you think this new issue is the same as or similar to #157 ?

Yes! Sorry I missed it. I scanned the issues earlier but somehow overlooked this one. Thanks!

**Comment by jukent** [comment]
@kmpaul Could you approve this? Your changes are still marked as "requested" even though they're resolved. Would love to merge this soon.

**Review by kmpaul — APPROVED** [review]
Looks good.  Thanks, @jukent!

**Pull request merged**